python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+            if timeout is not None:
+                timeout = max(0, timeout - (time.time() - start))
+
+        while self._next_chunk_index < len(
+                self._chunk_object_ids
+        ) and self._ready_indices[self._next_chunk_index]:
+            self._ready_objects.extend(
+                self._result_thread.results()[self._next_chunk_index])
+            self._next_chunk_index += 1
+
+        return self._ready_objects.pop(0)
+
+
+class UnorderedIMapIterator(IMapIterator):
+    def next(self, timeout=None):
+        if len(self._ready_objects) != 0:
+            return self._ready_objects.pop(0)
+
+        if self._next_chunk_index == len(self._chunk_object_ids):
+            raise StopIteration
+
+        try:
+            index = self._result_thread.next_ready_index(timeout=timeout)
+            self._submit_next_chunk()
+        except queue.Empty:
+            raise TimeoutError
+
+        self._ready_objects.extend(self._result_thread.results()[index])
+        self._next_chunk_index += 1
+
+        return self._ready_objects.pop(0)
+
+
+@ray.remote
+class PoolActor(object):
+    def __init__(self, initializer=None, initargs=None):
+        if initializer:
+            if initargs is None:
+                initargs = ()
+            initializer(*initargs)
+
+    def ping(self):
+        pass
+
+    def run_batch(self, func, batch):
+        results = []
+        for args, kwargs in batch:
+            if args is None:
+                args = tuple()
+            if kwargs is None:
+                kwargs = {}
+            try:
+                results.append(func(*args, **kwargs))
+            except Exception as e:
+                results.append(PoolTaskError(e))
+        return results
+
+
+# https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.pool
+# TODO(edoakes): cloudpickle can't pickle generator objects.
+class Pool(object):
+    def __init__(self,
+                 processes=None,
+                 initializer=None,
+                 initargs=None,
+                 maxtasksperchild=None):
+        self._closed = False
+        self._initializer = initializer
+        self._initargs = initargs
+        self._maxtasksperchild = maxtasksperchild if maxtasksperchild else -1
+        self._actor_deletion_ids = []
+
+        processes = self._init_ray(processes)
+        self._start_actor_pool(processes)
+
+    def _init_ray(self, processes=None):
+        if not ray.is_initialized():
+            # Cluster mode.
+            if "RAY_ADDRESS" in os.environ:
+                address = os.environ["RAY_ADDRESS"]
+                print("Connecting to ray cluster at address='{}'".format(
+                    address))
+                ray.init(address=address)
+            # Local mode.
+            else:
+                print("Starting local ray cluster")
+                ray.init(num_cpus=processes)
+
+        ray_cpus = int(ray.state.cluster_resources()["CPU"])
+        if processes is None:
```

Comment:
Can `processes` be zero? Do you want to check for it? It can result in confusing errors otherwise.